### ISSUE TITLE
Two issues when generating classes with embedded arrays of structs 

### ISSUE THREAD

**Issue by jpnoir** [mono/CppSharp#1598]
Target: C#. Will submit PR.

Consider the structs

```c++
struct DLL_API EmbeddedStruct
{
    uint64_t ui64;
};
struct DLL_API StructWithEmbeddedArrayOfStructObjectAlignment
{
    bool boolean;
    EmbeddedStruct embedded_struct[2];
};
```

*Issue 1:* With the existing implementation, `StructWithEmbeddedArrayOfStructObjectAlignment` is translated to

```c#
[StructLayout(LayoutKind.Sequential, Size = 24)]
public partial struct __Internal
{
    internal byte boolean;
    internal fixed byte embedded_struct[16];
    
    ...
}
```

Note that there is no padding following the `boolean` field. Since `embedded_struct` has been translated to a `fixed byte[]`, on the C# side, its offset is 1. On the C++ side, `embedded_struct` has an offset of 8. See `CSharpTests.TestAlignment` in the associated PR for a repro.

*Issue 2:* Observe that for the same struct(s), the accessor method for `Embeddedstruct` is

```c#
get
{
    global::CSharp.EmbeddedStruct[] __value = null;
    if (((__Internal*)__Instance)->embedded_struct != null)
    {
        __value = new global::CSharp.EmbeddedStruct[2];
        for (int i = 0; i < 2; i++)
            __value[i] =  global::CSharp.EmbeddedStruct.__CreateInstance(*((global::CSharp.EmbeddedStruct.__Internal*)&(((__Internal*)__Instance)->embedded_struct[i * sizeof(global::CSharp.EmbeddedStruct.__Internal)])));
    }
    return __value;
}
```

In particular, note that every time the accessor is called, it initializes a new instance of the `EmbeddedStruct[]` array by calling `__CreateInstance` using the overload that makes a copy of the relevant values in `embedded_struct[]` and creates a new instance of managed class, `EmbeddedStruct`, with that copy. Any updates to the array elements are written to the copy and not to `embedded_struct[16]`. The copied unmanaged memory is leaked. See `CSharpTests.TestEmbeddedArrayOfStructAccessor` in the associate PR for a repro.

##### Proposed fixes:

*Issue 1:* Change `LayoutFieldExtensions.CalculateOffset` to recognize the case of arrays of embedded structs and return the effective aligned offset that will be true when the array of structs is transformed to the fixed byte[] array so that appropriate padding will be generated by its caller, `CSharpSources.GenerateClassInternalsFields`.

*Issue 2:* Use `__GetOrCreateInstance` instead of `__CreateInstance` in the accessor. This results in returning the same managed instance each time and the first managed instance is initialized using the `_CreateInstance` overload which directly uses the native memory it is passed from the `embedded_struct[]` array.

**Comment by jpnoir** [comment]
@tritao, Not sure if you saw my response to your observation about the merged build failing since the PR is now marked closed. I'll repeat it here just in case:

I don't understand it either. The log message(s) indicate some sort of CI-time SDK configuration error.

I tried pulling the newly merged main branch and (of course) had no trouble restoring / building locally.

The projects that are failing are all C++/CLI projects which seems to match the docs pointed to by the URL in the error messages in the log. I haven't yet found where Visual Studio / MS Build establishes the expected AppHost version. However, if I'm interepreting the error message correctly, the version it found is (3.1.14). That version is included in the latest VS2019 version, v16.9.4. The previous Visual Studio version (v16.9.3) uses 3.1.13. The doc that the URL in the message refers you to suggests that scenario might be the issue:

> "This is typically caused by having a newer SDK than the one included in Visual Studio for C++/CLI projects. "

I can't see any version info for Visual Studio in the logs. Could this be the problem? Is it possible to update Visual Studio on the CI server?

**Comment by tritao** [comment]
Hello, first of all thanks for looking into this.

From what I understand, the CI environment is specified here: https://github.com/actions/virtual-environments/blob/main/images/win/Windows2019-Readme.md

From the CI log, we can also check the specific image version used by a specific run: https://github.com/mono/CppSharp/runs/2350856387?check_suite_focus=true#step:1:9

If we compare the run that failed and the one that worked:

Worked: `Version 20210404.2` 
Failed: `Version: 20210411.1`

I guess it's related. Maybe we need to change something on the Premake-generated project files to make them work with latest VS? I don't know but it's a possibility.

Can you check the contents of the C++/CLI vcxproj files, maybe there's some version we need to update?

**Comment by jpnoir** [comment]
@tritao, Thanks for the pointer to the environment description. 

Looking specifically at the VS2019 versions: GitHub is currently using:

> | Name                          | Version        |
> | ----------------------------- | -------------- |
> | Visual Studio Enterprise 2019 | 16.9.31129.286 |

and consulting the [magic decoder](https://docs.microsoft.com/en-us/visualstudio/install/visual-studio-build-numbers-and-release-dates?view=vs-2019): we see that the user-friendly version number for that Visual Studio is 16.9.3 which was released on March 30. As mentioned above, Visual Studio 16.9.3 expects to find pack version 3.1.13. The version we apparently have (based on the error message in the log) is 3.1.14. I observed in my own environment that when 3.1.14 got installed, 3.1.13 was removed.

When researching earlier, I found ways to hard-wire the expected pack SDK version, but I'm afraid that would break anyone else's environment who didn't have this inconsistency. Now that we've confirmed that we're on the right track. I'll look into workarounds a bit more.

**Comment by jpnoir** [comment]
@tritao, In doing some more experiments locally, I think I misinterpreted the error messages we're getting during the restore step: I think VS/msbuild is looking for AppHost 3.1.14 and couldn't find it. That seems to match up with the virtual environment description in the link you sent last week: it shows that we're running Microsoft .Net Core SDK v 3.1.407. The missing AppHost pack gets installed with SDK v3.1.408 (or, VS version 16.9.4). I still have no idea why msbuild is trying to find AppHost 3.1.14 if it's running 16.9.3.

I ran across this in the GitHub actions virtual environment issues section: https://github.com/actions/virtual-environments/issues/3200 where someone was requesting a VS update. The response seemed to indicate that the GitHub Actions team will be rolling out images with "Microsoft.NETCore.App 3.1.14" starting today. I'm hoping / guessing that means an update to the 3.1.408 SDK.

It's perhaps worth holding off for a few days to see if this update sorts things out.
